crowbar_framework/app/models/backup.rb

```diff
@@ -0,0 +1,186 @@
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+require "find"
+
+class Backup
+  include ActiveModel::Model
+
+  attr_accessor :name, :created_at, :filename, :path
+
+  validates_presence_of :name, :created_at
+  validates :created_at, exclusion: { in: [nil, ""] }
+  validate :filename, :filename_has_format
+  validate :filename, :filename_has_characters
+
+  def initialize(options)
+    @name = options.fetch :name, nil
+    @created_at = options.fetch :created_at, Time.zone.now.strftime("%Y%m%d-%H%M%S")
+    @filename = "#{@name}-#{@created_at}.tar.gz"
+    @path = Backup.image_dir.join(@filename)
+  end
+
+  def save
+    valid? && persist!
+  end
+
+  def exist?
+    !Backup.where(name: name, created_at: created_at).nil?
+  end
+
+  def delete
+    @path.delete
+  end
+
+  def size
+    self.path.file? ? self.path.size : nil
+  end
+
+  def upload(file)
+    Backup.image_dir.join(file.original_filename).open("wb") do |f|
+      f.write(file.read)
+    end
+  end
+
+  def restore
+    upgrade if upgrade?
+    ret = data_valid?
+    return ret unless ret[:status] == :ok
+    Crowbar::Backup::Restore.new(self).restore
+  end
+
+  def extract
+    backup_dir = Dir.mktmpdir
+    Archive.extract(path.to_s, backup_dir)
+    Pathname.new(backup_dir)
+  end
+
+  def data
+    @data ||= extract
+  end
+
+  def data_valid?
+    validate = Crowbar::Backup::Validate.new(data)
+    validate.validate
+  end
+
+  def version
+    @version ||= begin
+      data.join("crowbar", "version").read.strip
+    end
+  end
+
+  def upgrade?
+    ENV["CROWBAR_VERSION"].to_f > version.to_f
+  end
+
+  def upgrade
+    upgrade = Crowbar::Upgrade.new(self)
+    if upgrade.supported?
+      upgrade.upgrade
+    else
+      return {
+        status: :not_acceptable,
+        msg: I18n.t(
+          "backup.index.upgrade_not_supported"
+        )
+      }
+    end
+  end
+
+  class << self
+    def where(options = {})
+      name = options.fetch :name, nil
+      created_at = options.fetch :created_at, nil
+      ret = nil
+
+      all.each do |image|
+        ret = image if image.name == name && image.created_at == created_at
+      end
+      ret
+    end
+
+    def all
+      list = []
+
+      backup_files = image_dir.children.select do |c|
+        c.file? && c.to_s =~ /gz$/
+      end
+
+      backup_files.each do |backup_file|
+        name, created_at = filename_time(backup_file.basename.to_s)
+        image = new(
+          name: name,
+          created_at: created_at
+        )
+        list.push(image) if image.valid?
+      end
+      list.sort_by { |time| time.created_at }
```

Comment:
Style/SymbolProc: Pass `&:created_at` as an argument to `sort_by` instead of a block.